stdlib/public/core/Character.swift

```diff
@@ -245,21 +260,130 @@ extension Character
       "Can't form a Character from a String containing more than one extended grapheme cluster")
 
-    if _fastPath(s._guts.count <= 4) {
-      let b = _SmallUTF16(s._core)
-      if _fastPath(Int64(truncatingIfNeeded: b._storage) >= 0) {
-        _representation = .smallUTF16(
-          Builtin.trunc_Int64_Int63(b._storage._value))
+    self.init(_unverified: s._guts)
+  }
+
+  @_inlineable // FIXME(sil-serialize-all)
+  @_versioned // FIXME(sil-serialize-all)
+  internal init(_unverified guts: _StringGuts) {
+    defer { _fixLifetime(guts) }
+    if _slowPath(guts._isOpaque) {
+      let opaque = guts._asOpaque()
+      let native = opaque._copyToNativeStorage(of: UTF16.CodeUnit.self)
+      self.init(_unverified: native)
+    } else if guts.isASCII {
+      let ascii = guts._unmanagedASCIIView
+      if _fastPath(ascii.count == 1) {
+        self.init(_singleCodeUnit: ascii[0])
+      } else {
+        // The only multi-scalar ASCII grapheme cluster is CR/LF.
+        _sanityCheck(ascii.count == 2)
+        _sanityCheck(ascii.start[0] == _CR)
+        _sanityCheck(ascii.start[1] == _LF)
+        self.init(_codeUnitPair: UInt16(_CR), UInt16(_LF))
+      }
+    } else if guts._isNative {
+      self.init(_unverified: guts._nativeStorage(of: UTF16.CodeUnit.self))
+    } else {
+      self.init(_unverified: guts._unmanagedUTF16View)
+    }
+  }
+
+  /// Construct a Character from a slice of a _StringGuts, assuming
+  /// the specified range covers exactly one extended grapheme cluster.
+  @_inlineable // FIXME(sil-serialize-all)
+  @_versioned // FIXME(sil-serialize-all)
+  internal init(_unverified guts: _StringGuts, range: Range<Int>) {
+    defer { _fixLifetime(guts) }
+    if guts.isASCII {
+      let ascii = guts._unmanagedASCIIView
+      if _fastPath(range.count == 1) {
+        self.init(
+          _singleCodeUnit: ascii.codeUnit(atCheckedOffset: range.lowerBound))
+      } else {
+        // The only multi-scalar ASCII grapheme cluster is CR/LF.
+        ascii._boundsCheck(offsetRange: range)
+        _sanityCheck(range.count == 2)
+        _sanityCheck(ascii.start[range.lowerBound] == _CR)
+        _sanityCheck(ascii.start[range.lowerBound + 1] == _LF)
+        self.init(_codeUnitPair: UInt16(_CR), UInt16(_LF))
+      }
+    } else if _fastPath(range.count == 1) {
+      let cu = guts.codeUnit(atCheckedOffset: range.lowerBound)
+      // Check for half a surrogate pair. (We may encounter these at boundaries
+      // of String slices or when the data itself has an invalid encoding.)
+      if _fastPath(UTF16._isScalar(cu)) {
+        self.init(_singleCodeUnit: cu)
+      } else {
+        self.init(_singleCodeUnit: UTF16._replacementCodeUnit)
+      }
+    } else if guts._isContiguous {
+      let utf16 = guts._unmanagedUTF16View.checkedSlice(range)
+      self.init(_unverified: utf16)
+    } else {
+      let opaque = guts._asOpaque().checkedSlice(range)
+      self.init(_unverified: opaque._copyToNativeStorage())
+    }
+  }
```

Comment:
Just a general note, when it can aid readability, early returns and early exist are greatly encouraged.
http://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

Here, I think checking for a range.count == 1 first and early return would be more clear.